keps/sig-scheduling/4671-gang-scheduling/README.md

```diff
@@ -469,19 +519,269 @@ states (e.g. not yet block resources) will help with improving the scheduling ac
 Finally making the binding process aware of gangs will allow to make sure the process is either
 successful or triggers workload rescheduling satisfying requirement (7).
 
-The workload-aware preemption is tightly coupled, but separate feature that will also be designed
-in a dedicated KEP. The current vision includes introducing a dedicated preemption policy (that
-will result in pods no longer being treated individually for preemption purposes) which makes it
-an additive feature. However, having a next level of details is required to ensure that we really
-have a feasible backward-compatible plan before promoting this feature to Beta.
-
 Addressing requirement (8) is the biggest effort as it requires much closer integration between
 scheduler and autoscaling components. So in the initial steps we will only focus on mitigating
 this problem with existing mechanisms (e.g. reserving resources via NominatedNodeName).
 
 However, approval for this KEP is NOT an approval for this vision. We only sketch it to show that
 we see a viable path forward from the proposed design that will not require significant rework.
 
+### Scheduler Changes for Beta
+
+For the `Alpha` phase, we focused on plumbing the `Workload` API and implementing
+the `GangScheduling` plugin using simple barriers (`PreEnqueue` and `Permit`).
+While this satisfied the correctness requirement for "all-or-nothing" scheduling,
+it did not address performance or efficiency at scale, scheduling deadlocks,
+nor did it solve the problem of partial preemption application.
+
+For `Beta`, we propose introducing a **Workload Scheduling Cycle**.
+This mechanism processes all Pods belonging to a single `PodGroup` in one batch,
+rather than attempting to schedule them individually in isolation using the
+traditional pod-by-pod approach.
+While this won't fully address the "optimal enough" part of requirement (2),
+it ensures that all gang pods are processed together.
+The single scheduling cycle, together with blocking resources using nomination,
+will address requirement (3).
```

Comment:
In alpha, we avoided the case where two workloads each have some of their pods bound and some not bound, for an unbounded amount of time (deadlock).   However, it is possible that both workloads could both keep timing out before getting scheduled. (what might be called "livelock").

In beta, IIUC, avoid the "livelock" case by trying to do all the pods of one workload before starting work on another workload.

Reply:
+1